File: bindings/kotlin/src/main/kotlin/com/hivellm/umicp/discovery/ToolDiscovery.kt
package com.hivellm.umicp.discovery

import kotlinx.serialization.Serializable

/**
 * Tool Discovery for UMICP v0.2.0
 * MCP-compatible automatic tool introspection
 */

typealias JSONSchema = Map<String, Any>

/**
 * Operation schema compatible with MCP JSON Schema
 */
data class OperationSchema(
    val name: String,
    val inputSchema: JSONSchema,
    val title: String? = null,
    val description: String? = null,
    val outputSchema: JSONSchema? = null,
    val annotations: Map<String, Any?>? = null
)

/**
 * Server information for discovery
 */
data class ServerInfo(
    val server: String,
    val version: String,
    val protocol: String,
    val features: List<String>? = null,
    val operationsCount: Int? = null,
    val mcpCompatible: Boolean? = null,
    val metadata: Map<String, Any?>? = null
)

/**
 * Interface for services that support tool discovery
 */
interface DiscoverableService {
    /**
     * List all available operations with their schemas
     */
    fun listOperations(): List<OperationSchema>

    /**
     * Get schema for a specific operation by name
     */
    fun getSchema(name: String): OperationSchema? {
        return listOperations().find { it.name == name }
    }

    /**
     * Get server information and metadata
     */
    fun getServerInfo(): ServerInfo
}

/**
 * Builder for OperationSchema
 */
class OperationSchemaBuilder(
    private val name: String,
    private val inputSchema: JSONSchema
) {
    private var title: String? = null
    private var description: String? = null
    private var outputSchema: JSONSchema? = null
    private var annotations: Map<String, Any>? = null

    fun withTitle(title: String) = apply { this.title = title }
    fun withDescription(description: String) = apply { this.description = description }
    fun withOutputSchema(schema: JSONSchema) = apply { this.outputSchema = schema }
    fun withAnnotations(annotations: Map<String, Any?>) = apply { this.annotations = annotations }

    fun build() = OperationSchema(name, inputSchema, title, description, outputSchema, annotations)
}

/**
 * Builder for ServerInfo
 */
class ServerInfoBuilder(
    private val server: String,
    private val version: String,
    private val protocol: String
) {
    private var features: List<String>? = null
    private var operationsCount: Int? = null
    private var mcpCompatible: Boolean? = null
    private var metadata: Map<String, Any>? = null

    fun withFeatures(features: List<String>) = apply { this.features = features }
    fun withOperationsCount(count: Int) = apply { this.operationsCount = count }
    fun withMcpCompatible(compatible: Boolean) = apply { this.mcpCompatible = compatible }
    fun withMetadata(metadata: Map<String, Any?>) = apply { this.metadata = metadata }

    fun build() = ServerInfo(server, version, protocol, features, operationsCount, mcpCompatible, metadata)
}

/**
 * Helper functions for generating discovery responses
 */
object DiscoveryHelpers {
    /**
     * Generate JSON response for _list_operations
     */
    fun generateOperationsResponse(service: DiscoverableService): Map<String, Any> {
        val operations = service.listOperations()
        val info = service.getServerInfo()

        return mapOf(
            "operations" to operations,
            "count" to operations.size,
            "protocol" to info.protocol,
            "mcp_compatible" to (info.mcpCompatible ?: false)
        )
    }

    /**
     * Generate JSON response for _get_schema
     */
    fun generateSchemaResponse(service: DiscoverableService, operationName: String): Map<String, Any> {
        val schema = service.getSchema(operationName)

        return if (schema != null) {
            mapOf(
                "name" to schema.name,
                "title" to schema.title,
                "description" to schema.description,
                "input_schema" to schema.inputSchema,
                "output_schema" to schema.outputSchema,
                "annotations" to schema.annotations
            ).filterValues { it != null }
        } else {
            mapOf(
                "error" to "Operation not found",
                "operation" to operationName
            )
        }
    }

    /**
     * Generate JSON response for _server_info
     */
    fun generateServerInfoResponse(service: DiscoverableService): ServerInfo {
        return service.getServerInfo()
    }
}

/**
 * Simple implementation of DiscoverableService
 */
class SimpleDiscoverableService(
    private val operations: List<OperationSchema>,
    private val serverInfo: ServerInfo
) : DiscoverableService {

    override fun listOperations() = operations

    override fun getSchema(name: String) = operations.find { it.name == name }

    override fun getServerInfo() = serverInfo.copy(operationsCount = operations.size)
}

File: bindings/go/pkg/discovery/tool_discovery.go
package discovery

// Tool Discovery for UMICP v0.2.0
// MCP-compatible automatic tool introspection

// JSONSchema represents a JSON Schema (map with any structure)
type JSONSchema map[string]interface{}

// OperationSchema represents an operation with its input/output schemas
type OperationSchema struct {
	Name         string                 `json:"name"`
	Title        string                 `json:"title,omitempty"`
	Description  string                 `json:"description,omitempty"`
	InputSchema  JSONSchema             `json:"input_schema"`
	OutputSchema *JSONSchema            `json:"output_schema,omitempty"`
	Annotations  map[string]interface{} `json:"annotations,omitempty"`
}

// ServerInfo contains server metadata for discovery
type ServerInfo struct {
	Server          string                 `json:"server"`
	Version         string                 `json:"version"`
	Protocol        string                 `json:"protocol"`
	Features        []string               `json:"features,omitempty"`
	OperationsCount *int                   `json:"operations_count,omitempty"`
	McpCompatible   *bool                  `json:"mcp_compatible,omitempty"`
	Metadata        map[string]interface{} `json:"metadata,omitempty"`
}

// DiscoverableService interface for services that support tool discovery
type DiscoverableService interface {
	// ListOperations returns all available operations with their schemas
	ListOperations() []OperationSchema

	// GetSchema returns schema for a specific operation by name
	GetSchema(name string) *OperationSchema

	// GetServerInfo returns server information and metadata
	GetServerInfo() ServerInfo
}

// OperationSchemaBuilder provides a builder pattern for OperationSchema
type OperationSchemaBuilder struct {
	schema OperationSchema
}

// NewOperationSchema creates a new operation schema builder
func NewOperationSchema(name string, inputSchema JSONSchema) *OperationSchemaBuilder {
	return &OperationSchemaBuilder{
		schema: OperationSchema{
			Name:        name,
			InputSchema: inputSchema,
		},
	}
}

// WithTitle sets the title
func (b *OperationSchemaBuilder) WithTitle(title string) *OperationSchemaBuilder {
	b.schema.Title = title
	return b
}

// WithDescription sets the description
func (b *OperationSchemaBuilder) WithDescription(desc string) *OperationSchemaBuilder {
	b.schema.Description = desc
	return b
}

// WithOutputSchema sets the output schema
func (b *OperationSchemaBuilder) WithOutputSchema(schema JSONSchema) *OperationSchemaBuilder {
	b.schema.OutputSchema = &schema
	return b
}

// WithAnnotations sets annotations
func (b *OperationSchemaBuilder) WithAnnotations(annotations map[string]interface{}) *OperationSchemaBuilder {
	b.schema.Annotations = annotations
	return b
}

// Build returns the built OperationSchema
func (b *OperationSchemaBuilder) Build() OperationSchema {
	return b.schema
}

// ServerInfoBuilder provides a builder pattern for ServerInfo
type ServerInfoBuilder struct {
	info ServerInfo
}

// NewServerInfo creates a new server info builder
func NewServerInfo(server, version, protocol string) *ServerInfoBuilder {
	return &ServerInfoBuilder{
		info: ServerInfo{
			Server:   server,
			Version:  version,
			Protocol: protocol,
		},
	}
}

// WithFeatures sets the features
func (b *ServerInfoBuilder) WithFeatures(features []string) *ServerInfoBuilder {
	b.info.Features = features
	return b
}

// WithOperationsCount sets the operations count
func (b *ServerInfoBuilder) WithOperationsCount(count int) *ServerInfoBuilder {
	b.info.OperationsCount = &count
	return b
}

// WithMcpCompatible sets MCP compatibility
func (b *ServerInfoBuilder) WithMcpCompatible(compatible bool) *ServerInfoBuilder {
	b.info.McpCompatible = &compatible
	return b
}

// WithMetadata sets metadata
func (b *ServerInfoBuilder) WithMetadata(metadata map[string]interface{}) *ServerInfoBuilder {
	b.info.Metadata = metadata
	return b
}

// Build returns the built ServerInfo
func (b *ServerInfoBuilder) Build() ServerInfo {
	return b.info
}

// GenerateOperationsResponse generates a response for _list_operations
func GenerateOperationsResponse(service DiscoverableService) map[string]interface{} {
	operations := service.ListOperations()
	info := service.GetServerInfo()

	mcpCompatible := false
	if info.McpCompatible != nil {
		mcpCompatible = *info.McpCompatible
	}

	return map[string]interface{}{
		"operations":     operations,
		"count":          len(operations),
		"protocol":       info.Protocol,
		"mcp_compatible": mcpCompatible,
	}
}

// GenerateSchemaResponse generates a response for _get_schema
func GenerateSchemaResponse(service DiscoverableService, operationName string) map[string]interface{} {
	schema := service.GetSchema(operationName)

	if schema != nil {
		return map[string]interface{}{
			"name":          schema.Name,
			"title":         schema.Title,
			"description":   schema.Description,
			"input_schema":  schema.InputSchema,
			"output_schema": schema.OutputSchema,
			"annotations":   schema.Annotations,
		}
	}

	return map[string]interface{}{
		"error":     "Operation not found",
		"operation": operationName,
	}
}

// GenerateServerInfoResponse generates a response for _server_info
func GenerateServerInfoResponse(service DiscoverableService) ServerInfo {
	return service.GetServerInfo()
}

// CreateDiscoverableService creates a simple discoverable service from operations and info
func CreateDiscoverableService(operations []OperationSchema, serverInfo ServerInfo) DiscoverableService {
	count := len(operations)
	serverInfo.OperationsCount = &count

	return &simpleDiscoverableService{
		operations: operations,
		serverInfo: serverInfo,
	}
}

// simpleDiscoverableService is a simple implementation of DiscoverableService
type simpleDiscoverableService struct {
	operations []OperationSchema
	serverInfo ServerInfo
}

func (s *simpleDiscoverableService) ListOperations() []OperationSchema {
	return s.operations
}

func (s *simpleDiscoverableService) GetSchema(name string) *OperationSchema {
	for i := range s.operations {
		if s.operations[i].Name == name {
			return &s.operations[i]
		}
	}
	return nil
}

func (s *simpleDiscoverableService) GetServerInfo() ServerInfo {
	return s.serverInfo
}
